trieste/models/gpflux/architectures.py

```diff
@@ -0,0 +1,301 @@
+            centroids,
+            np.random.randn(num_inducing, latent_dim),
+        ],
+        axis=1,
+    )
+
+    for i_layer in range(num_layers):
+        is_first_layer = i_layer == 0
+        is_last_layer = i_layer == num_layers - 1
+        D_in = input_dim + latent_dim if is_first_layer else input_dim
+        D_out = 1 if is_last_layer else input_dim
+
+        z_init = centroids_ext if is_first_layer else centroids
+        inducing_var = construct_basic_inducing_variables(
+            num_inducing=num_inducing, input_dim=D_in, share_variables=True, z_init=z_init.copy()
+        )
+
+        kernel = construct_basic_kernel(
+            kernels=_construct_kernel(D_in, latent_dim, is_first_layer),
+            output_dim=D_out,
+            share_hyperparams=True,
+        )
+
+        if is_first_layer:
+            mean_function = gpflow.mean_functions.Zero()
+            q_sqrt_scaling = inner_layer_sqrt_factor
+        elif is_last_layer:
+            mean_function = gpflow.mean_functions.Zero()
+            q_sqrt_scaling = 1e-5
+        else:
+            mean_function = gpflow.mean_functions.Identity()
+            q_sqrt_scaling = inner_layer_sqrt_factor
+
+        layer = GPLayer(
+            kernel, inducing_var, num_data, mean_function=mean_function, name=f"gp_{i_layer}"
+        )
+        layer.q_sqrt.assign(layer.q_sqrt * q_sqrt_scaling)
+        gp_layers.append(layer)
+
+    likelihood = gpflow.likelihoods.Gaussian(likelihood_noise_variance)
+    return DeepGP(gp_layers, LikelihoodLayer(likelihood))
+
+
+class ModifiedLatentVariableLayer(LatentVariableLayer):
+    """
+    Modified version of :class:`gpflux.layers.LatentVariableLayer` to enable BayesOpt-specific
+    capabilities, such as variable dataset size.
+    """
+
+    def __init__(
+        self,
+        num_data: int,
+        prior: tfp.distributions.Distribution,
+        encoder: tf.keras.layers.Layer,
+        compositor: Optional[tf.keras.layers.Layer] = None,
+        name: Optional[str] = None,
+    ):
+        super().__init__(prior, encoder, compositor, name)
+
+        self.num_data = num_data
+
+    def _inference_latent_samples_and_loss(
+        self, layer_inputs: TensorType, observations: ObservationType, seed: Optional[int] = None
+    ) -> Tuple[tf.Tensor, tf.Tensor]:
+        distribution_params = self.encoder(None, training=True)
+        posteriors = self.distribution_class(*distribution_params, allow_nan_stats=False)
+        samples = posteriors.sample(seed=seed)[: tf.shape(layer_inputs)[0]]
+        local_kls = self._local_kls(posteriors)[: tf.shape(layer_inputs)[0]]
+        loss_per_datapoint = tf.reduce_mean(local_kls, name="local_kls")
+        return samples, loss_per_datapoint
+
+
+def _construct_kernel(
+    input_dim: int,
+    latent_dim: int,
+    is_first_layer: bool,
+) -> gpflow.kernels.SquaredExponential:
+    if is_first_layer:
+        lengthscales = [0.05] * (input_dim - latent_dim) + [2.0] * latent_dim
```

Comment:
where do the constants come from?

Reply:
These are just "sensible" initializations for the model, since we want to have more sensitivity to the data than the latent variables - if they're about the same scale then there's a risk the model will get stuck in a local minimum where it tries to explain all the data as noise.